src/shared/Core/CollectorWriter.cs

```diff
@@ -0,0 +1,172 @@
+            {
+                return;
+            }
+
+            Stop();
+
+            _pipeClient?.Dispose();
+            _queue.Dispose();
+
+            _pipeClient = null;
+            _writerThread = null;
+
+            _isDisposed = true;
+        }
+
+        public void Stop()
+        {
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(CollectorWriter));
+            }
+
+            if (_queue.IsAddingCompleted)
+            {
+                return;
+            }
+
+            // Signal to the queue draining thread that it should drain once more and then terminate.
+            _queue.CompleteAdding();
+            _writerThread.Join();
+
+            Debug.Assert(_queue.IsCompleted, "Message queue should be empty after being stopped");
+        }
+
+        public bool TryEnqueue(string message)
+        {
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(CollectorWriter));
+            }
+
+            return _queue.TryAdd(message);
+        }
+
+        private void BackgroundWriterThreadProc()
+        {
+            // Drain the queue of all messages currently in the queue.
+            // TryTake() using an infinite timeout will block until either a message is available (returns true)
+            // or the queue has been marked as completed _and_ is empty (returns false).
+            while (_queue.TryTake(out string message, Timeout.Infinite))
+            {
+                if (message != null)
+                {
+                    WriteMessage(message);
+                }
+            }
+
+            RaiseStateChanged(CollectorWriterState.Stopped, null);
+        }
+
+        private void RaiseStateChanged(CollectorWriterState newState, Exception ex)
+        {
+            State = newState;
+        }
+
+        private void WriteMessage(string message)
+        {
+            try
+            {
+                // If we're in byte/stream transmission mode rather than message mode
+                // we should signal the end of each message with a line-feed (LF) character.
+                if (_pipeClient.TransmissionMode == PipeTransmissionMode.Byte)
+                {
+                    message += '\n';
+                }
+
+                byte[] data = Encoding.UTF8.GetBytes(message);
+                _pipeClient.Write(data, 0, data.Length);
+                _pipeClient.Flush();
+
+                RaiseStateChanged(CollectorWriterState.Healthy, null);
+            }
+            catch (Exception ex)
+            {
+                // We can't send this message for some reason (e.g., broken pipe); we attempt no recovery or retry
+                // mechanism and drop this message. We will try to recreate/connect the pipe on the next message.
+                _pipeClient.Dispose();
+                _pipeClient = null;
+                RaiseStateChanged(CollectorWriterState.Failing, ex);
```

Comment:
If you get a SIGPIPE or something, just give up and turn off telemetry.  Don't try to restart it in the middle of the process.

GCM is only running for the duration of a single command, a missed message in the middle of the lifetime of this one invocation won't affect future invocations.  And the daemon cannot re-sync after an arbitrary dropped message -- for example, if you drop a region-enter, the daemon won't know what to do with the later (not dropped) region-leave.

If git.exe gets an error writing to the socket/pipe, it just turns off Trace2 and silently continues.
It's more important to not slow down the command by waiting to retry (or trying lots of re-connects).

If the daemon drops the connection (or sees that the client has dropped it), it is going to cleanup/flush the incomplete state for this client invocation anyway -- it won't try waiting around for the client to re-connect.

Reply:
I would just remove all of the state monitoring and events, and replace it with a simple "failed" flag that is set if the initial `Start/Connect` didn't work, or whenever an exception is thrown sending a message.

Maybe we should log to `Trace1` on failure? :)